File: src/dynamic_programming/cpp/count construct/test/test_count_construct.cpp
#include "count_construct.h"
#include "gtest/gtest.h"
#include <algorithm>

TEST(CountConstructTest, TestNegativeBasic1) {
  std::vector<std::string> wordBank = {"bo",  "rd", "ate", "t",
                                       "ska", "sk", "boar"};
  std::string target = "skateboard";
  auto result = 0;

  EXPECT_EQ(result, countConstructBasic(target, wordBank));
}

TEST(CountConstructTest, TestNegativeBasic2) {
  std::vector<std::string> wordBank = {"mo", "ha", "cz"};
  std::string target = "mohach";
  auto result = 0;

  EXPECT_EQ(result, countConstructBasic(target, wordBank));
}

TEST(CountConstructTest, TestPositiveBasic1) {
  std::vector<std::string> wordBank = {"a", "b", "c"};
  std::string target = "abc";
  auto result = 1;

  EXPECT_EQ(result, countConstructBasic(target, wordBank));
}

TEST(CountConstructTest, TestPositiveBasic2) {
  std::vector<std::string> wordBank = {"ab", "abc", "cd", "def", "abcd"};
  std::string target = "abcdef";
  auto result = 1;

  EXPECT_EQ(result, countConstructBasic(target, wordBank));
}

TEST(CountConstructTest, TestNegativeMemo1) {
  std::vector<std::string> wordBank = {"bo",  "rd", "ate", "t",
                                       "ska", "sk", "boar"};
  std::string target = "skateboard";
  auto result = 0;

  EXPECT_EQ(result, countConstructMemo(target, wordBank));
}

TEST(CountConstructTest, TestNegativeMemo2) {
  std::vector<std::string> wordBank = {"mo", "ha", "cz"};
  std::string target = "mohach";
  auto result = 0;

  EXPECT_EQ(result, countConstructMemo(target, wordBank));
}

TEST(CountConstructTest, TestPositiveMemo1) {
  std::vector<std::string> wordBank = {"a", "b", "c"};
  std::string target = "abc";
  auto result = 1;

  EXPECT_EQ(result, countConstructMemo(target, wordBank));
}

TEST(CountConstructTest, TestPositiveMemo2) {
  std::vector<std::string> wordBank = {"ab", "abc", "cd", "def", "abcd"};
  std::string target = "abcdef";
  auto result = 1;

  EXPECT_EQ(result, countConstructMemo(target, wordBank));
}

TEST(CountConstructTest, TestNegativeTable1) {
  std::vector<std::string> wordBank = {"bo",  "rd", "ate", "t",
                                       "ska", "sk", "boar"};
  std::string target = "skateboard";
  auto result = 0;

  EXPECT_EQ(result, countConstructTable(target, wordBank));
}

TEST(CountConstructTest, TestNegativeTable2) {
  std::vector<std::string> wordBank = {"mo", "ha", "cz"};
  std::string target = "mohach";
  auto result = 0;

  EXPECT_EQ(result, countConstructTable(target, wordBank));
}

TEST(CountConstructTest, TestPositiveTable1) {
  std::vector<std::string> wordBank = {"a", "b", "c"};
  std::string target = "abc";
  auto result = 1;

  EXPECT_EQ(result, countConstructTable(target, wordBank));
}

TEST(CountConstructTest, TestPositiveTable2) {
  std::vector<std::string> wordBank = {"ab", "abc", "cd", "def", "abcd"};
  std::string target = "abcdef";
  auto result = 1;

  EXPECT_EQ(result, countConstructTable(target, wordBank));
}

File: src/dynamic_programming/python/count_construct/tests/test_count_construct.py
import unittest


from src.count_construct import (
    count_construct_basic,
    count_construct_memo,
    count_construct_table,
)


class TestCountConstructBasic(unittest.TestCase):
    def test_negative_1(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        result = 0
        self.assertEqual(count_construct_basic(target, word_bank), result)

    def test_negative_2(self):
        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        result = 0
        self.assertEqual(count_construct_basic(target, word_bank), result)

    def test_positive_1(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        result = 1
        self.assertEqual(count_construct_basic(target, word_bank), result)

    def test_positive_2(self):
        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        result = 1
        self.assertEqual(count_construct_basic(target, word_bank), result)


class TestCountConstructMemo(unittest.TestCase):
    def test_negative_1(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        result = 0
        self.assertEqual(count_construct_memo(target, word_bank), result)

    def test_negative_2(self):
        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        result = 0
        self.assertEqual(count_construct_memo(target, word_bank), result)

    def test_positive_1(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        result = 1
        self.assertEqual(count_construct_memo(target, word_bank), result)

    def test_positive_2(self):
        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        result = 1
        self.assertEqual(count_construct_memo(target, word_bank), result)


class TestCountConstructTab(unittest.TestCase):
    def test_negative_1(self):
        word_bank = ["bo", "rd", "ate", "t", "ska", "sk", "boar"]
        target = "skateboard"
        result = 0
        self.assertEqual(count_construct_table(target, word_bank), result)

    def test_negative_2(self):
        word_bank = ["mo", "ha", "cz"]
        target = "mocha"
        result = 0
        self.assertEqual(count_construct_table(target, word_bank), result)

    def test_positive_1(self):
        word_bank = ["a", "b", "c"]
        target = "abc"
        result = 1
        self.assertEqual(count_construct_table(target, word_bank), result)

    def test_positive_2(self):
        word_bank = ["ab", "abc", "cd", "def", "abcd"]
        target = "abcdef"
        result = 1
        self.assertEqual(count_construct_table(target, word_bank), result)


if __name__ == "__main__":
    unittest.main()
